File: data/onboardmeasurements/Scripts/Implied_Field_Correlation/CalculateImpliedField.py
import pandas as pd
import numpy as np
import os
import glob
import re
import matplotlib.pyplot as plt

def process_onboard_measurements(target_board="Peters_Board"):
    # Base directories relative to script or repo root
    script_dir = os.path.dirname(os.path.abspath(__file__))
    repo_root = os.path.abspath(os.path.join(script_dir, "..", "..", ".."))
    
    base_onboard = os.path.join(repo_root, "data", "onboardmeasurements", target_board)
    base_response = os.path.join(repo_root, "data", "responsemeasurements", "LibraryBased", "Week5", target_board)
    
    if not os.path.exists(base_onboard):
        base_onboard = os.path.join(script_dir, "..", "..", target_board)
        base_response = os.path.join(script_dir, "..", "..", "..", "responsemeasurements", "LibraryBased", "Week5", target_board)
        
    # Directory to save the correlation plots
    analysis_dir = os.path.join(base_onboard, "Analysis", "ImpliedField")
    os.makedirs(analysis_dir, exist_ok=True)
    
    print("======================================================")
    print(f"   ONBOARD VS RESPONSE CORRELATION ({target_board.upper()})")
    print("======================================================\n")

    # Find all onboard Excel files
    onboard_files = glob.glob(os.path.join(base_onboard, 'Board*', '**', '*.xlsx'), recursive=True)
    
    results_summary = []
    
    for file in sorted(onboard_files):
        # Ignore temporary lock files and anything already in Analysis
        if '~$' in file or 'Analysis' in file:
            continue
            
        try:
            rel = os.path.relpath(file, base_onboard)
            parts = os.path.normpath(rel).split(os.sep)
            board_folder = parts[0]
            filename = parts[-1]
            orientation = parts[1] if len(parts) > 2 else 'Default'
            
            # Locate response measurement file
            response_file = os.path.join(base_response, board_folder, filename)
            
            # Fallback 1: check if orientation folder exists in response measurements
            if not os.path.isfile(response_file) and orientation != 'Default':
                cand = os.path.join(base_response, board_folder, orientation, filename)
                if os.path.isfile(cand):
                    response_file = cand
                    
            # Fallback 2: check if filename has _X or -X variants
            if not os.path.isfile(response_file):
                if '_X' in filename:
                    alt1 = os.path.join(base_response, board_folder, filename.replace('_X', ''))
                    alt2 = os.path.join(base_response, board_folder, filename.replace('_X', '-X'))
                    if os.path.isfile(alt1): response_file = alt1
                    elif os.path.isfile(alt2): response_file = alt2
                elif '-X' in filename:
                    alt1 = os.path.join(base_response, board_folder, filename.replace('-X', ''))
                    alt2 = os.path.join(base_response, board_folder, filename.replace('-X', '_X'))
                    if os.path.isfile(alt1): response_file = alt1
                    elif os.path.isfile(alt2): response_file = alt2
                    
            # Fallback 3: check if file was saved in an unexpected board folder (e.g. B4A under Board2)
            if not os.path.isfile(response_file):
                m = re.match(r'B(\d+)', filename)
                if m:
                    correct_board = f'Board{m.group(1)}'
                    cand = os.path.join(base_response, correct_board, filename)
                    if os.path.isfile(cand):
                        response_file = cand
                    elif '_X' in filename:
                        alt1 = os.path.join(base_response, correct_board, filename.replace('_X', ''))
                        if os.path.isfile(alt1): response_file = alt1
                        
            if not os.path.isfile(response_file):
                print(f"Skipping {filename} in {board_folder}/{orientation}: No corresponding response measurement found.")
                continue
                
            # 1. Read Response Measurement for the true Sensitivity
            resp_df = pd.read_excel(response_file)
            
            # Extract Sensitivity (Ohms/G)
            if 'Sensitivity_Ohms_per_G' in resp_df.columns:
                sens = resp_df['Sensitivity_Ohms_per_G'].iloc[0]
            else:
                sens = resp_df.iloc[:, 4].iloc[0]
                
            # 2. Read Onboard Measurement
            onboard_df = pd.read_excel(file)
            
            if 'Resistance_Ohms' in onboard_df.columns:
                R = onboard_df['Resistance_Ohms']
                I = onboard_df['Kepco_Current_A']
            else:
                I = onboard_df.iloc[:, 0]
                R = onboard_df.iloc[:, 1]
                
            # 3. Calculate Implied Gauss
            # Find the resistance when current is 0A to use as our 0 Gauss generated field baseline
            idx_0 = np.argmin(np.abs(I))
            R_at_0A = R.iloc[idx_0] if isinstance(R, pd.Series) else R[idx_0]
            implied_G = (R - R_at_0A) / sens
            
            # 4. Correlate Implied Gauss with the Onboard Current
            # Since magnetic field strength is proportional to current magnitude, we correlate Implied G with absolute Current (A)
            max_I = np.abs(I).max()
            max_G = implied_G.max()
            
            # Calculate the linear correlation slope (Gauss per Ampere)
            slope, intercept = np.polyfit(I, implied_G, 1)
            
            # Calculate RMSE of the linear fit
            implied_G_fit = slope * I + intercept
            rmse = np.sqrt(np.mean((implied_G - implied_G_fit)**2))
            
            # Print the results nicely
            rel_path = os.path.relpath(file, base_onboard)
            print(f"File: {rel_path}")
            print(f"  -> True Sensitivity (from Response): {sens:.4f} Ohms/G")
            print(f"  -> Max Implied Field (Generated):    {max_G:.4f} G")
            print(f"  -> Max Coil Current:                 {max_I:.4f} A")
            print(f"  -> Generated Field Correlation:      {slope:.4f} G/A")
            print(f"  -> RMSE (Fit Error):                 {rmse:.4f} G")
            print("-" * 54)
            
            results_summary.append({
                'Board': board_folder,
                'Orientation': orientation,
                'File': filename,
                'Sensitivity_Ohms_per_G': sens,
                'Max_Implied_G': max_G,
                'Max_Current_A': max_I,
                'Correlation_G_per_A': slope,
                'RMSE_G': rmse
            })
            
            # 5. Save a plot of the Implied Field vs Current
            plt.figure(figsize=(8, 5))
            
            # Split forward and backward sweep
            max_idx = np.argmax(I.values if hasattr(I, 'values') else I)
            if 0 < max_idx < len(I) - 1:
                plt.plot(I[:max_idx+1], implied_G[:max_idx+1], 'b-', linewidth=2.0, label='Forward Sweep')
                plt.plot(I[max_idx:], implied_G[max_idx:], 'r-', linewidth=2.0, label='Backward Sweep')
            else:
                plt.plot(I, implied_G, 'b-', linewidth=2.0, label='Sweep')
                
            clean_filename = filename.replace('.xlsx', '')
            plt.title(f"Implied Magnetic Field vs Onboard Current\n{board_folder} - {clean_filename} ({orientation})")
            plt.xlabel("Kepco Current (A)")
            plt.ylabel("Implied Magnetic Field (G)")
            plt.grid(True)
            plt.legend()
            
            plot_name = f"{board_folder}_{orientation}_{clean_filename}.png"
            plt.tight_layout()
            plt.savefig(os.path.join(analysis_dir, plot_name))
            plt.close()
            
        except Exception as e:
            print(f"Error processing {os.path.relpath(file, base_onboard)}: {e}")

    # Save summary excel sheet
    if results_summary:
        df_summary = pd.DataFrame(results_summary)
        summary_path = os.path.join(analysis_dir, 'Correlation_Summary.xlsx')
        df_summary.to_excel(summary_path, index=False)
        print(f"Saved summary to {summary_path}")

if __name__ == "__main__":
    process_onboard_measurements("Peters_Board")

File: data/onboardmeasurements/Scripts/Field_Correlation/CalculateImpliedField.m
% CalculateImpliedField.m
% This script correlates the onboard measurements with the true sensitivity 
% derived from the response measurements. It calculates the Implied Magnetic 
% Field (Gauss) and finds the linear correlation (Gauss per Ampere) generated 
% by the onboard coil.

% =========================================================================
% CONFIGURATION: Select Target Board
% =========================================================================
% Set target board: 'Peters_Board' or 'Hsin_Tings_Board'
target_board = 'Peters_Board'; 

% Determine script directory reliably
script_dir = fileparts(mfilename('fullpath'));
if isempty(script_dir)
    script_dir = pwd;
end

% Define base directories relative to this script's location
base_onboard = fullfile(script_dir, '..', '..', target_board); 
base_response = fullfile(script_dir, '..', '..', '..', 'responsemeasurements', 'LibraryBased', 'Week5', target_board);

% Fallback checks if script is run from different working directories (e.g. repo root or board folder)
if ~exist(base_onboard, 'dir')
    if exist(fullfile(pwd, 'data', 'onboardmeasurements', target_board), 'dir')
        base_onboard = fullfile(pwd, 'data', 'onboardmeasurements', target_board);
        base_response = fullfile(pwd, 'data', 'responsemeasurements', 'LibraryBased', 'Week5', target_board);
    elseif exist(fullfile(pwd, target_board), 'dir')
        base_onboard = fullfile(pwd, target_board);
        base_response = fullfile(pwd, '..', '..', 'responsemeasurements', 'LibraryBased', 'Week5', target_board);
    elseif exist(fullfile(pwd, '..', target_board), 'dir')
        base_onboard = fullfile(pwd, '..', target_board);
        base_response = fullfile(pwd, '..', '..', '..', 'responsemeasurements', 'LibraryBased', 'Week5', target_board);
    elseif exist(fullfile(script_dir, 'Board1'), 'dir')
        base_onboard = script_dir;
        base_response = fullfile(script_dir, '..', '..', 'responsemeasurements', 'LibraryBased', 'Week5', target_board);
    end
end

% Create directory to save the correlation plots
analysis_dir = fullfile(base_onboard, 'Analysis', 'ImpliedField');
if ~exist(analysis_dir, 'dir')
    mkdir(analysis_dir);
end

fprintf('======================================================\n');
fprintf('   ONBOARD VS RESPONSE CORRELATION (%s)\n', upper(strrep(target_board, '_', ' ')));
fprintf('======================================================\n\n');

% Find all onboard Excel files in all subdirectories of Board*
files = dir(fullfile(base_onboard, 'Board*', '**', '*.xlsx'));

% Initialize results cell array for Excel summary
results_summary = {};

for i = 1:length(files)
    file_path = fullfile(files(i).folder, files(i).name);
    
    % Ignore temporary lock files and anything already inside Analysis
    if contains(file_path, '~$') || contains(file_path, 'Analysis')
        continue;
    end
    
    try
        % Extract folder parts dynamically
        parts = strsplit(file_path, filesep);
        filename = parts{end};
        
        % Locate Board* directory in path
        board_idx = -1;
        for j = 1:length(parts)
            if startsWith(parts{j}, 'Board') && length(parts{j}) <= 6
                board_idx = j;
                break;
            end
        end
        
        if board_idx == -1
            continue;
        end
        
        board_folder = parts{board_idx};
        if board_idx == length(parts) - 1
            orientation = 'Default';
        else
            orientation = parts{board_idx + 1};
        end
        
        % Locate the exact same board and test point in response measurements
        response_file = fullfile(base_response, board_folder, filename);
        
        % Fallback 1: check if orientation folder exists in response measurements
        if ~isfile(response_file) && ~strcmp(orientation, 'Default')
            cand = fullfile(base_response, board_folder, orientation, filename);
            if isfile(cand)
                response_file = cand;
            end
        end
        
        % Fallback 2: check if filename has _X or -X variants
        if ~isfile(response_file)
            if contains(filename, '_X')
                alt1 = fullfile(base_response, board_folder, strrep(filename, '_X', ''));
                alt2 = fullfile(base_response, board_folder, strrep(filename, '_X', '-X'));
                if isfile(alt1)
                    response_file = alt1;
                elseif isfile(alt2)
                    response_file = alt2;
                end
            elseif contains(filename, '-X')
                alt1 = fullfile(base_response, board_folder, strrep(filename, '-X', ''));
                alt2 = fullfile(base_response, board_folder, strrep(filename, '-X', '_X'));
                if isfile(alt1)
                    response_file = alt1;
                elseif isfile(alt2)
                    response_file = alt2;
                end
            end
        end
        
        % Fallback 3: check if file was saved in an unexpected board folder (e.g. B4A under Board2)
        if ~isfile(response_file)
            if startsWith(filename, 'B') && length(filename) >= 2 && isstrprop(filename(2), 'digit')
                correct_board = ['Board' filename(2)];
                cand = fullfile(base_response, correct_board, filename);
                if isfile(cand)
                    response_file = cand;
                elseif contains(filename, '_X')
                    cand_alt = fullfile(base_response, correct_board, strrep(filename, '_X', ''));
                    if isfile(cand_alt)
                        response_file = cand_alt;
                    end
                end
            end
        end
        
        if ~isfile(response_file)
            fprintf('Skipping %s in %s/%s: No corresponding response measurement found.\n', filename, board_folder, orientation);
            continue;
        end
        
        % 1. Read Response Measurement for the true Sensitivity
        opts = detectImportOptions(response_file);
        resp_df = readtable(response_file, opts);
        
        % Extract Sensitivity (Ohms/G)
        if ismember('Sensitivity_Ohms_per_G', resp_df.Properties.VariableNames)
            sens = resp_df.Sensitivity_Ohms_per_G(1);
        else
            sens = resp_df{1, 5}; % fallback to 5th column
        end
        
        % 2. Read Onboard Measurement
        opts_onboard = detectImportOptions(file_path);
        onboard_df = readtable(file_path, opts_onboard);
        
        if ismember('Resistance_Ohms', onboard_df.Properties.VariableNames)
            R = onboard_df.Resistance_Ohms;
            I = onboard_df.Kepco_Current_A;
        else
            I = onboard_df{:, 1};
            R = onboard_df{:, 2};
        end
        
        % 3. Calculate Implied Gauss
        % Find the resistance when current is 0A to use as our 0 Gauss baseline
        [~, idx_0] = min(abs(I));
        R_at_0A = R(idx_0);
        implied_G = (R - R_at_0A) / sens;
        
        % 4. Correlate Implied Gauss with the Onboard Current
        % Since magnetic field strength is proportional to current magnitude, we correlate Implied G with absolute Current (A)
        max_I = max(abs(I));
        max_G = max(implied_G);
        
        % Calculate the linear correlation slope (Gauss per Ampere)
        p = polyfit(I, implied_G, 1);
        slope = p(1);
        
        % Print the results nicely
        rel_path = fullfile(board_folder, orientation, filename);
        fprintf('File: %s\n', rel_path);
        fprintf('  -> True Sensitivity (from Response): %.4f Ohms/G\n', sens);
        fprintf('  -> Max Implied Field (Generated):    %.4f G\n', max_G);
        fprintf('  -> Max Coil Current:                 %.4f A\n', max_I);
        fprintf('  -> Generated Field Correlation:      %.4f G/A\n', slope);
        fprintf('------------------------------------------------------\n');
        
        % Add to summary
        results_summary(end+1, :) = {board_folder, orientation, filename, sens, max_G, max_I, slope};
        
        % 5. Save a plot of the Implied Field vs Current
        fig = figure('Visible', 'off');
        set(fig, 'Units', 'normalized', 'Position', [0.2 0.2 0.4500 0.400]);
        
        % Split forward and backward sweep
        [~, max_idx] = max(I);
        if max_idx > 1 && max_idx < length(I)
            plot(I(1:max_idx), implied_G(1:max_idx), 'b-', 'LineWidth', 2.0, 'DisplayName', 'Forward Sweep');
            hold on;
            plot(I(max_idx:end), implied_G(max_idx:end), 'r-', 'LineWidth', 2.0, 'DisplayName', 'Backward Sweep');
            hold off;
        else
            plot(I, implied_G, 'b-', 'LineWidth', 2.0, 'DisplayName', 'Sweep');
        end
        
        % Clean up the title for display
        clean_filename = strrep(filename, '.xlsx', '');
        title_str = sprintf('Implied Magnetic Field vs Onboard Current\n%s - %s (%s)', board_folder, clean_filename, orientation);
        title(title_str, 'Interpreter', 'none', 'FontSize', 16, 'FontWeight', 'bold');
        
        xlabel('Kepco Current (A)', 'FontSize', 14, 'FontWeight', 'bold');
        ylabel('Implied Magnetic Field (G)', 'FontSize', 14, 'FontWeight', 'bold');
        
        ax = gca;
        ax.LineWidth = 1.5;
        ax.Box = 'on';
        
        lgd = legend('Location', 'best');
        lgd.FontSize = 12;
        lgd.FontWeight = 'bold';
        grid on;
        
        plot_name = sprintf('%s_%s_%s.fig', board_folder, orientation, clean_filename);
        plot_png = sprintf('%s_%s_%s.png', board_folder, orientation, clean_filename);
        
        % Set figure visible right before saving so it opens visibly when opened in MATLAB
        set(fig, 'Visible', 'on');
        savefig(fig, fullfile(analysis_dir, plot_name));
        saveas(fig, fullfile(analysis_dir, plot_png));
        close(fig);
        
    catch ME
        fprintf('Error processing %s: %s\n', file_path, ME.message);
    end
end

% Save summary to Excel
if ~isempty(results_summary)
    T = cell2table(results_summary, 'VariableNames', {'Board', 'Orientation', 'File', 'Sensitivity_Ohms_per_G', 'Max_Implied_G', 'Max_Current_A', 'Correlation_G_per_A'});
    summary_path = fullfile(analysis_dir, 'Correlation_Summary.xlsx');
    writetable(T, summary_path);
    fprintf('Saved summary to %s\n', summary_path);
end
